src/transformers/models/cogvlm/modeling_cogvlm.py

```diff
@@ -0,0 +1,1145 @@
+
+        return (
+            (attn_output, attn_weights, past_key_value) if output_attentions else (attn_output, None, past_key_value)
+        )
+
+
+class CogvlmVisionExpertSdpaAttention(CogvlmVisionExpertAttention):
+    """
+    CogVLM attention module using torch.nn.functional.scaled_dot_product_attention. This module inherits from
+    `CogvlmVisionExpertAttention` as the weights of the module stays untouched. The only changes are on the forward pass to adapt to
+    SDPA API.
+    """
+
+    # Adapted from CogvlmVisionExpertAttention.forward
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        token_type_ids: torch.LongTensor,
+        position_ids: torch.LongTensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        past_key_value: Optional[Cache] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        batch_size, q_len, hidden_size = hidden_states.size()
+        vision_token_mask, language_token_mask = get_expert_mask(token_type_ids)
+
+        mixed_raw_layer = torch.empty(
+            (batch_size, q_len, hidden_size * 3), dtype=hidden_states.dtype, device=hidden_states.device
+        )
+        mixed_raw_layer[vision_token_mask] = self.vision_expert_query_key_value(hidden_states[vision_token_mask])
+        mixed_raw_layer[language_token_mask] = self.language_expert_query_key_value(hidden_states[language_token_mask])
+
+        query_states, key_states, value_states = torch.split(mixed_raw_layer, self.hidden_size, dim=-1)
+        query_states = self._transpose_for_scores(query_states)
+        key_states = self._transpose_for_scores(key_states)
+        value_states = self._transpose_for_scores(value_states)
+
+        kv_seq_len = key_states.shape[-2]
+        if past_key_value is not None:
+            kv_seq_len += past_key_value.get_seq_length(self.layer_idx)
+
+        cos, sin = self.rotary_emb(value_states, seq_len=position_ids.max() + 1)
+        cos, sin = (
+            nn.functional.embedding(position_ids, cos.squeeze(1)),
+            nn.functional.embedding(position_ids, sin.squeeze(1)),
+        )
+        query_states, key_states = apply_rotary_pos_emb(
+            query_states, key_states, cos, sin, position_ids, unsqueeze_dim=1
+        )
+
+        if past_key_value is not None:
+            cache_kwargs = {"sin": sin, "cos": cos}  # Specific to RoPE models
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
+
+        attention_mask_bool = attention_mask == 0
+        is_full = (attention_mask_bool > 0).all()
+
+        context_layer = torch.nn.functional.scaled_dot_product_attention(
+            query_states,
+            key_states,
+            value_states,
+            attn_mask=None,
+            dropout_p=0.0,
+            is_causal=not is_full,
```

Comment:
What if `is_full` is False, but we in fact don't have a causal mask? Can't that happen?

Also, aren't we impacted by SDPA causal mask layout https://github.com/pytorch/pytorch/issues/108108?